File: SimplePlayer/PlayerWindowController.m
/*
 * Copyright (c) 2009 - 2020 Stephen F. Booth <me@sbooth.org>
 *
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "PlayerWindowController.h"

#import <SFBAudioEngine/SFBAudioDecoder.h>
#import <SFBAudioEngine/SFBAudioFile.h>
#import <SFBAudioEngine/SFBAudioPlayer.h>

@interface SFBAttachedPicture (ImageCreation)
@property (nonatomic, nullable, readonly) NSImage *image;
@end

@implementation SFBAttachedPicture (ImageCreation)
- (NSImage *)image { return [[NSImage alloc] initWithData:self.imageData]; }
@end

@interface PlayerWindowController ()
{
@private
	SFBAudioPlayer 		*_player;
	dispatch_source_t	_timer;
}
@end

@interface PlayerWindowController (Private)
- (void) updateWindowUI;
@end

@implementation PlayerWindowController

- (id) init
{
	if((self = [super initWithWindowNibName:@"PlayerWindow"])) {
		_player = [[SFBAudioPlayer alloc] init];

		__weak typeof(self) weakSelf = self;

		[_player setRenderingStartedNotificationHandler:^(SFBAudioDecoder * _Nonnull decoder) {
			NSURL *url = decoder.inputSource.url;
			dispatch_async(dispatch_get_main_queue(), ^{
				[weakSelf updateWindowUI];
				if([url isFileURL])
					[[NSDocumentController sharedDocumentController] noteNewRecentDocumentURL:url];
			});
		}];

		[_player setRenderingFinishedNotificationHandler:^(SFBAudioDecoder * _Nonnull decoder) {
#pragma unused(decoder)
			dispatch_async(dispatch_get_main_queue(), ^{
				[weakSelf updateWindowUI];
			});
		}];

		[_player setErrorNotificationHandler:^(NSError * _Nonnull error) {
			dispatch_async(dispatch_get_main_queue(), ^{
				[NSApp presentError:error];
			});
		}];

		// Update the UI 5 times per second
		_timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_main_queue());
		dispatch_source_set_timer(_timer, DISPATCH_TIME_NOW, NSEC_PER_SEC / 5, NSEC_PER_SEC / 3);

		dispatch_source_set_event_handler(_timer, ^{
			if(!self->_player.isPlaying)
				[self.playButton setTitle:@"Resume"];
			else
				[self.playButton setTitle:@"Pause"];

			SFBAudioPlayerPlaybackPosition playbackPosition;
			SFBAudioPlayerPlaybackTime playbackTime;
			if([self->_player getPlaybackPosition:&playbackPosition andTime:&playbackTime]) {
				double fractionComplete = (double)playbackPosition.currentFrame / playbackPosition.totalFrames;
				[self.slider setDoubleValue:fractionComplete];
				[self.elapsed setDoubleValue:playbackTime.currentTime];
				[self.remaining setDoubleValue:(-1 * (playbackTime.totalTime - playbackTime.currentTime))];
			}
		});

		// Start the timer
		dispatch_resume(_timer);
	}

	return self;
}

- (void) awakeFromNib
{
	// Disable the UI since no file is loaded
	[self updateWindowUI];
}

- (NSString *) windowFrameAutosaveName
{
	return @"Player Window";
}

- (void) windowWillClose:(NSNotification *)notification
{
#pragma unused(notification)
	[_player stopReturningError:nil];
}

- (IBAction) playPause:(id)sender
{
#pragma unused(sender)
	[_player playPauseReturningError:nil];
}

- (IBAction) seekForward:(id)sender
{
#pragma unused(sender)
	[_player seekForward];
}

- (IBAction) seekBackward:(id)sender
{
#pragma unused(sender)
	[_player seekBackward];
}

- (IBAction) seek:(id)sender
{
#pragma unused(sender)
	[_player seekToPosition:[sender floatValue]];
}

- (IBAction) skipToNextTrack:(id)sender
{
#pragma unused(sender)
	[_player skipToNext];
}

- (BOOL)playURL:(NSURL *)url
{
	return [_player playURL:url error:nil];
}

- (BOOL)enqueueURL:(NSURL *)url
{
	return [_player enqueueURL:url error:nil];
}

@end

@implementation PlayerWindowController (Private)

- (void)updateWindowUI
{
	NSURL *url = _player.url;

	// Nothing happening, reset the window
	if(!url) {
		[[self window] setRepresentedURL:nil];
		[[self window] setTitle:@""];

		[self.slider setEnabled:NO];
		[self.playButton setState:NSOffState];
		[self.playButton setEnabled:NO];
		[self.backwardButton setEnabled:NO];
		[self.forwardButton setEnabled:NO];

		[self.elapsed setHidden:YES];
		[self.remaining setHidden:YES];

		[self.albumArt setImage:[NSImage imageNamed:@"NSApplicationIcon"]];
		[self.title setStringValue:@""];
		[self.artist setStringValue:@""];

		return;
	}

	BOOL seekable = _player.supportsSeeking;

	// Update the window's title and represented file
	[[self window] setRepresentedURL:url];
	[[self window] setTitle:[[NSFileManager defaultManager] displayNameAtPath:[url path]]];

	[[NSDocumentController sharedDocumentController] noteNewRecentDocumentURL:url];

	// Update the UI
	[self.slider setEnabled:seekable];
	[self.playButton setEnabled:YES];
	[self.backwardButton setEnabled:seekable];
	[self.forwardButton setEnabled:seekable];

	// Show the times
	[self.elapsed setHidden:NO];

	if(_player.totalFrames != -1)
		[self.remaining setHidden:NO];

	// Load and display some metadata.  Normally the metadata would be read and stored in the background,
	// but for simplicity's sake it is done here.
	SFBAudioFile *audioFile = [SFBAudioFile audioFileWithURL:url error:nil];
	if(audioFile) {
		SFBAudioMetadata *metadata = audioFile.metadata;
		NSSet *pictures = metadata.attachedPictures;
		if([pictures count] > 0)
			[self.albumArt setImage:[[pictures anyObject] image]];
		else
			[self.albumArt setImage:nil];

		[self.title setStringValue:metadata.title ?: @""];
		[self.artist setStringValue:metadata.artist ?: @""];
	}
	else {
		[self.albumArt setImage:[NSImage imageNamed:@"NSApplicationIcon"]];
		[self.title setStringValue:@""];
		[self.artist setStringValue:@""];
	}
}

@end

File: SimplePlayer/PlayerWindowController.swift
/*
 * Copyright (c) 2009 - 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

import Cocoa

extension AttachedPicture {
	var image: NSImage? {
		return NSImage(data: imageData)
	}
}

class PlayerWindowController: NSWindowController, NSWindowDelegate {
	var player: AudioPlayer!
	var timer: DispatchSourceTimer!

	@IBOutlet weak var slider: NSSlider!
	@IBOutlet weak var elapsed: NSTextField!
	@IBOutlet weak var remaining: NSTextField!
	@IBOutlet weak var playButton: NSButton!
	@IBOutlet weak var forwardButton: NSButton!
	@IBOutlet weak var backwardButton: NSButton!
	@IBOutlet weak var albumArt: NSImageView!
	@IBOutlet weak var title: NSTextField!
	@IBOutlet weak var artist: NSTextField!

	override var windowNibName: NSNib.Name? {
		return "PlayerWindow"
	}

	override func windowDidLoad() {
		player = AudioPlayer()

		player.renderingStartedNotificationHandler = { decoder in
			let url = decoder.inputSource.url
			DispatchQueue.main.async {
				self.updateWindow()
				if let url = url, url.isFileURL {
					NSDocumentController.shared.noteNewRecentDocumentURL(url)
				}
			}
		}

		player.renderingFinishedNotificationHandler = { decoder in
			DispatchQueue.main.async {
				self.updateWindow()
			}
		}

		player.decodingCanceledNotificationHandler = { decoder in
			if !self.player.isRunning {
				DispatchQueue.main.async {
					self.updateWindow()
				}
			}
		}

		timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
		timer.schedule(deadline: DispatchTime.now(), repeating: .milliseconds(200), leeway: .milliseconds(100))

		timer.setEventHandler {
			if self.player.isPlaying {
				self.playButton.title = "Pause"
			}
			else {
				self.playButton.title = "Resume"
			}

			let positionAndTime = self.player.positionAndTime

			if positionAndTime.position.current != -1 && positionAndTime.position.total != -1 {
				self.slider.doubleValue = Double(positionAndTime.position.current) / Double(positionAndTime.position.total)
			}

			if positionAndTime.time.current != -1 {
				self.elapsed.doubleValue = positionAndTime.time.current
				if positionAndTime.time.total != -1 {
					self.remaining.doubleValue = -1 * (positionAndTime.time.total - positionAndTime.time.current)
				}
			}
		}

		timer.resume()

		updateWindow()
	}

	func windowWillClose(_ notification: Notification) {
		player.stop()
	}

	@IBAction func playPause(_ sender: AnyObject?) {
		try? player.playPause()
	}

	@IBAction func seekForward(_ sender: AnyObject?) {
		player.seekForward()
	}

	@IBAction func seekBackward(_ sender: AnyObject?) {
		player.seekBackward()
	}

	@IBAction func seek(_ sender: AnyObject?) {
		if let position = sender?.floatValue {
			player.seek(position: position)
		}
	}

	@IBAction func skipToNextTrack(_ sender: AnyObject?) {
		player.skipToNext()
	}

	lazy var supportedPathExtensions: [String] = {
		var pathExtensions = [String]()
		pathExtensions.append(contentsOf: AudioDecoder.supportedPathExtensions)
		pathExtensions.append(contentsOf: DSDDecoder.supportedPathExtensions)
		return pathExtensions
	}()

	private func decoder(for url: URL, enableDoP: Bool = false) throws -> PCMDecoding? {
		let pathExtension = url.pathExtension.lowercased()
		if AudioDecoder.handlesPaths(withExtension: pathExtension) {
			return try AudioDecoder(url: url)
		}
		else if DSDDecoder.handlesPaths(withExtension: pathExtension) {
			let dsdDecoder = try DSDDecoder(url: url)
			return enableDoP ? try DoPDecoder(decoder: dsdDecoder) : try DSDPCMDecoder(decoder: dsdDecoder)
		}
		return nil
	}

	func play(_ url: URL) {
		do {
			if let decoder = try decoder(for: url) {
				try player.play(decoder)
			}
		}
		catch let error {
			NSApp.presentError(error)
		}
	}

	func enqueue(_ url: URL) {
		do {
			if let decoder = try decoder(for: url) {
				try player.enqueue(decoder)
			}
		}
		catch let error {
			NSApp.presentError(error)
		}
	}

	func updateWindow() {
		// Nothing happening, reset the window
		guard let url = player.url else {
			window?.representedURL = nil
			window?.title = ""

			slider.isEnabled = false
			playButton.state = .off
			playButton.isEnabled = false
			backwardButton.isEnabled = false
			forwardButton.isEnabled = false

			elapsed.isHidden = true
			remaining.isHidden = true

			albumArt.image = NSImage(named: "NSApplicationIcon")
			title.stringValue = ""
			artist.stringValue = ""

			return
		}

		let seekable = player.supportsSeeking

		// Update the window's title and represented file
		window?.representedURL = url
		window?.title = FileManager.default.displayName(atPath: url.path)

		NSDocumentController.shared.noteNewRecentDocumentURL(url)

		// Update the UI
		slider.isEnabled = seekable
		playButton.state = .off
		playButton.isEnabled = true
		backwardButton.isEnabled = seekable
		forwardButton.isEnabled = seekable

		// Show the times
		elapsed.isHidden = false
		if player.frameLength != -1 {
			remaining.isHidden = false
		}

		// Load and display some metadata.  Normally the metadata would be read and stored in the background,
		// but for simplicity's sake it is done here.
		if let audioFile = try? AudioFile(readingPropertiesAndMetadataFrom: url) {
			let metadata = audioFile.metadata

			if let picture = metadata.attachedPictures.randomElement() {
				albumArt.image = picture.image
			}
			else {
				albumArt.image = nil
			}

			title.stringValue = metadata.title ?? ""
			artist.stringValue = metadata.artist ?? ""
		}
		else {
			albumArt.image = NSImage(named: "NSApplicationIcon")
			title.stringValue = ""
			artist.stringValue = ""
		}
	}
}
